File: src/llm/ollama.py
"""Ollama provider for local LLM inference."""

import asyncio

import httpx

from .base import BaseLLMProvider, Message, CompletionResult


class OllamaProvider(BaseLLMProvider):
    """LLM provider using Ollama for local inference.

    Ollama supports various open models like llama3, mistral, etc.
    Great for fully private, offline operation.
    """

    def __init__(
        self,
        base_url: str = "http://localhost:11434",
        model: str = "qwen3.5:4b",
        max_tokens: int = 300,
        timeout: float = 120.0,
        think: bool = False,
        keep_alive: str = "30m",
    ):
        """Initialize Ollama provider.

        Args:
            base_url: Ollama server URL
            model: Model to use (e.g., "qwen3.5:4b", "llama3.3", "mistral")
            max_tokens: Maximum tokens in response (Ollama uses num_predict)
            timeout: Request timeout in seconds
            think: Enable thinking/reasoning mode (slower, off by default)
            keep_alive: How long Ollama keeps the model in memory after a
                request. A generous finite window (not -1) so the model
                self-frees if the app crashes; graceful shutdown calls
                ``unload()`` to free it sooner. Resets on every request, so
                during active use the model effectively stays warm.
        """
        super().__init__(model=model, max_tokens=max_tokens)
        self.base_url = base_url.rstrip("/")
        self.timeout = timeout
        self.think = think
        self.keep_alive = keep_alive
        self._client: httpx.AsyncClient | None = None
        self._client_loop: asyncio.AbstractEventLoop | None = None

    async def _get_client(self) -> httpx.AsyncClient:
        """Get or create HTTP client.

        Recreates the client when the event loop changes (e.g. between
        successive ``asyncio.run()`` calls in the SocketIO handlers).
        """
        loop = asyncio.get_running_loop()
        if self._client is not None and self._client_loop is not loop:
            try:
                await self._client.aclose()
            except Exception:
                pass
            self._client = None
        if self._client is None:
            self._client = httpx.AsyncClient(
                timeout=httpx.Timeout(self.timeout, connect=5.0),
            )
            self._client_loop = loop
        return self._client

    async def complete(
        self,
        messages: list[Message],
        system: str | None = None,
        max_tokens: int | None = None,
    ) -> CompletionResult:
        """Generate a completion using Ollama."""
        client = await self._get_client()

        # Build messages list
        ollama_messages = []

        if system:
            ollama_messages.append({
                "role": "system",
                "content": system,
            })

        for msg in messages:
            ollama_messages.append({
                "role": msg.role,
                "content": msg.content,
            })

        # Make request — disable thinking by default for faster responses
        response = await client.post(
            f"{self.base_url}/api/chat",
            json={
                "model": self.model,
                "messages": ollama_messages,
                "stream": False,
                "think": self.think,
                "keep_alive": self.keep_alive,
                "options": {
                    "num_predict": max_tokens or self.max_tokens,
                },
            },
        )
        response.raise_for_status()

        data = response.json()

        # Extract response
        text = data.get("message", {}).get("content", "")

        # Ollama provides some usage info
        tokens_used = None
        input_tokens = output_tokens = None
        if "eval_count" in data:
            input_tokens = data.get("prompt_eval_count", 0)
            output_tokens = data.get("eval_count", 0)
            tokens_used = input_tokens + output_tokens

        return CompletionResult(
            text=text,
            finish_reason=data.get("done_reason"),
            tokens_used=tokens_used,
            input_tokens=input_tokens,
            output_tokens=output_tokens,
        )

    async def check_model_available(self) -> bool:
        """Check if the configured model is available.

        Returns:
            True if model is available
        """
        try:
            client = await self._get_client()
            response = await client.get(f"{self.base_url}/api/tags")
            response.raise_for_status()

            data = response.json()
            models = [m["name"] for m in data.get("models", [])]

            # Check for exact match or prefix match (e.g., "llama3" matches "llama3:latest")
            return any(
                m == self.model or m.startswith(f"{self.model}:")
                for m in models
            )
        except Exception:
            return False

    async def unload(self) -> None:
        """Ask Ollama to free the model from memory now (keep_alive=0).

        Best-effort: called on graceful session-end / app-close so the model
        doesn't linger for the full keep_alive window. If it fails (Ollama
        already down, network blip), the model still self-unloads after
        keep_alive, so we swallow errors.
        """
        try:
            client = await self._get_client()
            await client.post(
                f"{self.base_url}/api/chat",
                json={"model": self.model, "messages": [], "keep_alive": 0},
            )
        except Exception:
            pass

    async def close(self) -> None:
        """Close the HTTP client."""
        if self._client:
            await self._client.aclose()
            self._client = None


# ---------------------------------------------------------------------------
# Provider registry — maps provider names to their configuration.
#
# Each entry contains:
#   module:        relative import path for the provider class
#   class_name:    class to import from that module
#   default_model: used when the caller doesn't specify a model
#   kwargs_fn:     callable(factory_args) -> dict of constructor kwargs
#
# kwargs_fn receives the full set of factory keyword arguments so each
# provider can pick the ones it needs and apply its own defaults.
# ---------------------------------------------------------------------------

PROVIDERS: dict[str, dict] = {
    "claude_proxy": {
        "module": ".claude_proxy",
        "class_name": "ClaudeProxyProvider",
        "default_model": "sonnet",
        "kwargs_fn": lambda a: {},
    },
    "anthropic": {
        "module": ".anthropic",
        "class_name": "AnthropicProvider",
        "default_model": "claude-sonnet-4-6",
        "kwargs_fn": lambda a: {
            "api_key": a["api_key"],
        },
    },
    "openai": {
        "module": ".openai",
        "class_name": "OpenAIProvider",
        "default_model": "gpt-5.4-mini",
        "kwargs_fn": lambda a: {
            "api_key": a["api_key"],
            "base_url": a["base_url"],
        },
    },
    "openrouter": {
        "module": ".openai",
        "class_name": "OpenAIProvider",
        "default_model": "deepseek/deepseek-v3.2",
        "kwargs_fn": lambda a: {
            "api_key": a["api_key"],
            "base_url": "https://openrouter.ai/api/v1",
            "env_key": "OPENROUTER_API_KEY",
        },
    },
    "venice": {
        "module": ".openai",
        "class_name": "OpenAIProvider",
        "default_model": "llama-3.3-70b",
        "kwargs_fn": lambda a: {
            "api_key": a["api_key"],
            "base_url": "https://api.venice.ai/api/v1",
            "env_key": "VENICE_API_KEY",
            "extra_body": {"venice_parameters": {"include_venice_system_prompt": False}},
        },
    },
    "groq": {
        "module": ".openai",
        "class_name": "OpenAIProvider",
        "default_model": "llama-3.3-70b-versatile",
        "kwargs_fn": lambda a: {
            "api_key": a["api_key"],
            "base_url": "https://api.groq.com/openai/v1",
            "env_key": "GROQ_API_KEY",
        },
    },
    "ollama": {
        "module": None,  # OllamaProvider is defined in this file
        "class_name": "OllamaProvider",
        "default_model": "qwen3.5:4b",
        "kwargs_fn": lambda a: {
            "base_url": a["ollama_url"] or "http://localhost:11434",
        },
    },
}


def create_llm_provider(
    provider: str,
    model: str | None = None,
    ollama_url: str | None = None,
    api_key: str | None = None,
    max_tokens: int = 300,
    base_url: str | None = None,
) -> BaseLLMProvider:
    """Factory function to create LLM provider.

    Args:
        provider: Provider name ("claude_proxy", "anthropic", "openai", "ollama", "openrouter", "venice", "groq")
        model: Model name (uses provider default if not specified)
        ollama_url: Ollama server URL (for ollama)
        api_key: API key (for anthropic/openai/openrouter)
        max_tokens: Maximum response tokens
        base_url: Custom base URL for OpenAI-compatible APIs

    Returns:
        LLM provider instance
    """
    import importlib

    cfg = PROVIDERS.get(provider)
    if cfg is None:
        raise ValueError(f"Unknown LLM provider: {provider}")

    # Resolve the provider class via lazy import (or from this module)
    if cfg["module"] is not None:
        mod = importlib.import_module(cfg["module"], package=__package__)
        cls = getattr(mod, cfg["class_name"])
    else:
        cls = globals()[cfg["class_name"]]

    # Build kwargs: start with provider-specific ones, then add common ones
    factory_args = {
        "ollama_url": ollama_url,
        "api_key": api_key,
        "base_url": base_url,
    }
    kwargs = cfg["kwargs_fn"](factory_args)
    kwargs["model"] = model or cfg["default_model"]
    kwargs["max_tokens"] = max_tokens

    return cls(**kwargs)

File: ts/src/llm/ollama.ts
/**
 * Ollama provider — local inference via the Ollama HTTP API.
 *
 * Direct fetch implementation; no SDK dependency.
 */

import type {
    CompletionOptions,
    CompletionResult,
    LLMProvider,
    Message,
    StreamChunk,
} from './base.js';

const DEFAULT_BASE_URL = 'http://localhost:11434';
const DEFAULT_MODEL = 'qwen3.5:4b';
const DEFAULT_MAX_TOKENS = 300;

export interface OllamaProviderOptions {
    baseUrl?: string;
    model?: string;
    maxTokens?: number;
    /** Enable thinking/reasoning mode (slower, off by default). */
    think?: boolean;
    /** Override fetch for testing. */
    fetchImpl?: typeof fetch;
}

interface OllamaChatResponse {
    message?: { content?: string };
    done_reason?: string | null;
    prompt_eval_count?: number;
    eval_count?: number;
}

interface OllamaTagsResponse {
    models?: Array<{ name: string }>;
}

export class OllamaProvider implements LLMProvider {
    readonly model: string;
    readonly maxTokens: number;
    readonly think: boolean;
    private readonly baseUrl: string;
    private readonly fetchImpl: typeof fetch;

    constructor(options: OllamaProviderOptions = {}) {
        this.baseUrl = (options.baseUrl ?? DEFAULT_BASE_URL).replace(/\/+$/, '');
        this.model = options.model ?? DEFAULT_MODEL;
        this.maxTokens = options.maxTokens ?? DEFAULT_MAX_TOKENS;
        this.think = options.think ?? false;
        this.fetchImpl = options.fetchImpl ?? globalThis.fetch.bind(globalThis);
    }

    private buildBody(messages: Message[], options: CompletionOptions, stream: boolean): string {
        const ollamaMessages: Array<{ role: string; content: string }> = [];
        if (options.system) {
            ollamaMessages.push({ role: 'system', content: options.system });
        }
        for (const msg of messages) {
            ollamaMessages.push({ role: msg.role, content: msg.content });
        }
        return JSON.stringify({
            model: this.model,
            messages: ollamaMessages,
            stream,
            think: this.think,
            options: { num_predict: options.maxTokens ?? this.maxTokens },
        });
    }

    async complete(messages: Message[], options: CompletionOptions = {}): Promise<CompletionResult> {
        const response = await this.fetchImpl(`${this.baseUrl}/api/chat`, {
            method: 'POST',
            headers: { 'content-type': 'application/json' },
            body: this.buildBody(messages, options, false),
        });

        if (!response.ok) {
            const detail = await response.text().catch(() => '');
            throw new Error(`Ollama error ${response.status}: ${detail}`);
        }

        const data = (await response.json()) as OllamaChatResponse;
        const text = data.message?.content ?? '';

        return {
            text,
            finishReason: data.done_reason ?? null,
            ...ollamaUsage(data),
        };
    }

    async *completeStream(
        messages: Message[],
        options: CompletionOptions = {}
    ): AsyncIterable<StreamChunk> {
        const response = await this.fetchImpl(`${this.baseUrl}/api/chat`, {
            method: 'POST',
            headers: { 'content-type': 'application/json' },
            body: this.buildBody(messages, options, true),
        });

        if (!response.ok) {
            const detail = await response.text().catch(() => '');
            throw new Error(`Ollama error ${response.status}: ${detail}`);
        }
        if (!response.body) {
            throw new Error('Ollama streaming response has no body');
        }

        // Ollama streams NDJSON — one JSON object per line, each like:
        //   {"message":{"content":"Hello"},"done":false}
        //   {"message":{"content":""},"done":true,"eval_count":...}
        const reader = response.body.getReader();
        const decoder = new TextDecoder('utf-8');
        let buffer = '';
        let finishReason: string | null = null;
        let usage = { tokensUsed: null, inputTokens: null, outputTokens: null } as ReturnType<
            typeof ollamaUsage
        >;

        try {
            while (true) {
                const { value, done } = await reader.read();
                if (done) break;
                buffer += decoder.decode(value, { stream: true });
                let nl: number;
                while ((nl = buffer.indexOf('\n')) >= 0) {
                    const line = buffer.slice(0, nl).trim();
                    buffer = buffer.slice(nl + 1);
                    if (!line) continue;
                    const parsed = safeJson<OllamaStreamChunk>(line);
                    if (!parsed) continue;
                    const text = parsed.message?.content ?? '';
                    if (text.length > 0) {
                        yield { text, done: false };
                    }
                    if (parsed.done) {
                        finishReason = parsed.done_reason ?? 'stop';
                        usage = ollamaUsage(parsed);
                    }
                }
            }
        } finally {
            try {
                reader.releaseLock();
            } catch {
                /* ignore */
            }
        }

        yield { text: '', done: true, finishReason, ...usage };
    }

    /** True if the configured model (exact or prefix match) is pulled. */
    async checkModelAvailable(): Promise<boolean> {
        try {
            const response = await this.fetchImpl(`${this.baseUrl}/api/tags`);
            if (!response.ok) return false;
            const data = (await response.json()) as OllamaTagsResponse;
            const names = data.models?.map((m) => m.name) ?? [];
            return names.some((n) => n === this.model || n.startsWith(`${this.model}:`));
        } catch {
            return false;
        }
    }

    /**
     * If the configured model isn't currently loaded in Ollama's memory,
     * return a user-facing status string explaining the upcoming cold-load
     * wait. Returns null when the model is already loaded, when Ollama
     * isn't reachable, or when we can't determine load state — in all of
     * those cases the caller has nothing useful to show.
     *
     * Cheap (one HTTP call, 2s timeout); fine to call before every
     * completion. After first use the model stays loaded so subsequent
     * checks return null and the status banner clears itself.
     */
    async coldLoadMessage(): Promise<string | null> {
        try {
            const response = await this.fetchImpl(`${this.baseUrl}/api/ps`);
            if (!response.ok) return null;
            const data = (await response.json()) as OllamaPsResponse;
            const loaded = data.models?.map((m) => m.name) ?? [];
            const isLoaded = loaded.some(
                (n) => n === this.model || n.startsWith(`${this.model}:`)
            );
            if (isLoaded) return null;
            return `Loading ${this.model} into memory… first response can take a few seconds.`;
        } catch {
            return null;
        }
    }
}

interface OllamaPsResponse {
    models?: Array<{ name: string }>;
}

interface OllamaStreamChunk {
    message?: { content?: string };
    done?: boolean;
    done_reason?: string | null;
    prompt_eval_count?: number;
    eval_count?: number;
}

/**
 * Map Ollama's eval counts to the CompletionResult split: prompt_eval_count
 * is input, eval_count is output. Local models have no prompt caching, so no
 * cache fields. `tokensUsed` is the sum, null when the model reported nothing.
 */
function ollamaUsage(data: { prompt_eval_count?: number; eval_count?: number }): {
    tokensUsed: number | null;
    inputTokens: number | null;
    outputTokens: number | null;
} {
    if (data.eval_count === undefined && data.prompt_eval_count === undefined) {
        return { tokensUsed: null, inputTokens: null, outputTokens: null };
    }
    const inputTokens = data.prompt_eval_count ?? 0;
    const outputTokens = data.eval_count ?? 0;
    return { tokensUsed: inputTokens + outputTokens, inputTokens, outputTokens };
}

function safeJson<T>(s: string): T | null {
    try {
        return JSON.parse(s) as T;
    } catch {
        return null;
    }
}
